dbms/src/Interpreters/MergeJoin.cpp

```diff
@@ -320,21 +326,127 @@ void joinInequalsLeft(const Block & left_block, MutableColumns & left_columns, M
     appendNulls(right_columns, rows_to_add);
 }
 
+Blocks blocksListToBlocks(const BlocksList & in_blocks)
+{
+    Blocks out_blocks;
+    out_blocks.reserve(in_blocks.size());
+    for (const auto & block : in_blocks)
+        out_blocks.push_back(block);
+    return out_blocks;
+}
+
+std::unique_ptr<TemporaryFile> flushBlockToFile(const String & tmp_path, const Block & header, Block && block)
+{
+    auto tmp_file = createTemporaryFile(tmp_path);
+
+    OneBlockInputStream stream(block);
+    std::atomic<bool> is_cancelled{false};
+    TemporaryFileStream::write(tmp_file->path(), header, stream, &is_cancelled);
+    if (is_cancelled)
+        throw Exception("Cannot flush MergeJoin data on disk. No space at " + tmp_path, ErrorCodes::NOT_ENOUGH_SPACE);
+
+    return tmp_file;
+}
+
+void flushStreamToFiles(const String & tmp_path, const Block & header, IBlockInputStream & stream,
+                        std::vector<std::unique_ptr<TemporaryFile>> & files,
+                        std::function<void(const Block &)> callback = [](const Block &){})
+{
+    while (Block block = stream.read())
+    {
+        if (!block.rows())
+            continue;
+
+        callback(block);
+        auto tmp_file = flushBlockToFile(tmp_path, header, std::move(block));
+        files.emplace_back(std::move(tmp_file));
+    }
+}
+
+BlockInputStreams makeSortedInputStreams(std::vector<MiniLSM::SortedFiles> & sorted_files, const Block & header)
+{
+    BlockInputStreams inputs;
+
+    for (const auto & track : sorted_files)
+    {
+        BlockInputStreams sequence;
+        for (const auto & file : track)
+            sequence.emplace_back(std::make_shared<TemporaryFileLazyInputStream>(file->path(), header));
+        inputs.emplace_back(std::make_shared<ConcatBlockInputStream>(sequence));
+    }
+
+    return inputs;
 }
 
+}
 
 
-MergeJoin::MergeJoin(std::shared_ptr<AnalyzedJoin> table_join_, const Block & right_sample_block)
+void MiniLSM::insert(const BlocksList & blocks)
+{
+    if (blocks.empty())
+        return;
+
+    SortedFiles sorted_blocks;
+    if (blocks.size() > 1)
+    {
+        BlockInputStreams inputs;
+        inputs.reserve(blocks.size());
+        for (auto & block : blocks)
+            inputs.push_back(std::make_shared<OneBlockInputStream>(block));
+
+        MergingSortedBlockInputStream sorted_input(inputs, sort_description, rows_in_block);
+        flushStreamToFiles(path, sample_block, sorted_input, sorted_blocks);
+    }
+    else
+    {
+        OneBlockInputStream sorted_input(blocks.front());
+        flushStreamToFiles(path, sample_block, sorted_input, sorted_blocks);
+    }
+
+    sorted_files.emplace_back(std::move(sorted_blocks));
+    if (sorted_files.size() >= max_size)
+        merge();
+}
+
+/// TODO: better merge strategy
+void MiniLSM::merge(std::function<void(const Block &)> callback)
```

Comment:
Can we have a more descriptive name for the callback, like `on_before_flush_block`?